server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java

```diff
@@ -1130,6 +1144,110 @@ public void testWatermarksEnabledForSingleDataNode() {
                 " actual free: [20.0%]"));
     }
 
+    public void testDiskThresholdWithSnapshotShardSizes() {
+        final long shardSizeInBytes = randomBoolean() ? 10L : 50L;
+        logger.info("--> using shard size [{}]", shardSizeInBytes);
+
+        final Settings diskSettings = Settings.builder()
+            .put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_THRESHOLD_ENABLED_SETTING.getKey(), true)
+            .put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING.getKey(), "90%")
+            .put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING.getKey(), "95%")
+            .build();
+
+        final ImmutableOpenMap.Builder<String, DiskUsage> usagesBuilder = ImmutableOpenMap.builder();
+        usagesBuilder.put("node1", new DiskUsage("node1", "n1", "/dev/null", 100, 21));  // 79% used
+        usagesBuilder.put("node2", new DiskUsage("node2", "n2", "/dev/null", 100, 1)); // 99% used
+        final ImmutableOpenMap<String, DiskUsage> usages = usagesBuilder.build();
+        final ClusterInfoService clusterInfoService = () -> new DevNullClusterInfo(usages, usages, ImmutableOpenMap.of());
+
+        final AllocationDeciders deciders = new AllocationDeciders(
+            new HashSet<>(Arrays.asList(
+                new RestoreInProgressAllocationDecider(),
+                new SameShardAllocationDecider(
+                    Settings.EMPTY, new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS)
+                ),
+                makeDecider(diskSettings))));
+
+        final Snapshot snapshot = new Snapshot("_repository", new SnapshotId("_snapshot_uuid", "_snapshot_name"));
+        final IndexId indexId = new IndexId("_indexid_uuid", "_indexid_name");
+        final ShardId shardId = new ShardId(new Index("test", IndexMetadata.INDEX_UUID_NA_VALUE), 0);
+
+        final AtomicReference<SnapshotShardSizeInfo> snapshotShardSizeInfoRef = new AtomicReference<>(SnapshotShardSizeInfo.EMPTY);
+        final SnapshotsInfoService snapshotsInfoService = snapshotShardSizeInfoRef::get;
+
+        final AllocationService strategy = new AllocationService(deciders, new TestGatewayAllocator(),
```

Comment:
NIT: I think I would have had an easier time following this test if we had moved this to where it's actually first used after the assertions on the CS we construct below to like:

```java
        final AtomicReference<SnapshotShardSizeInfo> snapshotShardSizeInfoRef = new AtomicReference<>(SnapshotShardSizeInfo.EMPTY);
        final AllocationService strategy = new AllocationService(deciders, new TestGatewayAllocator(),
                new BalancedShardsAllocator(Settings.EMPTY), clusterInfoService, snapshotShardSizeInfoRef::get);

        // reroute triggers snapshot shard size fetching
        clusterState = strategy.reroute(clusterState, "reroute");
        logShardStates(clusterState);
```